src/azure-cli-core/azure/cli/core/commands/progress.py

```diff
@@ -0,0 +1,176 @@
+    Both = 2
+
+
+class ProgressViewBase(object):
+    """ a view base for progress reporting """
+    def __init__(self, out, progress_type, format_percent=None):
+        self.out = out
+        assert isinstance(progress_type, ProgressType)
+        self.progress_type = progress_type
+        self.format_percent = format_percent
+
+    def write(self, args):
+        """ writes the progress """
+        pass
+
+    def flush(self):
+        """ flushes the message out the pipeline"""
+        self.out.flush()
+
+    def get_type(self):
+        """ returns the progress type """
+        return self.progress_type
+
+
+class DetProgressReporter(object):
+    """ generic progress reporter """
+    def __init__(self, total_value=1):
+        self.message = ''
+        self.curr_val = 0 if total_value else None
+        self.total_val = total_value
+
+    def add(self, kwargs):
+        """
+        adds a progress report
+        :param kwargs: dictionary containing 'message', 'total_val', 'value'
+        """
+        message = kwargs.get('message', '')
+        total_val = kwargs.get('total_val', 1.0)
+        value = kwargs.get('value', 0.0)
+        assert value >= 0 and value <= total_val and total_val >= 0
+        self.total_val = total_val
+        self.curr_val = value
+        self.message = message
+
+    def report(self):
+        """ report the progress """
+        percent = self.curr_val / self.total_val
+        return {'message': self.message, 'percent': percent}
+
+
+class InDetProgressReporter(object):
+    """ generic progress reporter """
+    def __init__(self):
+        self.message = ''
+
+    def add(self, kwargs):
+        """ adds a progress report """
+        self.message = kwargs.get('message', '')
+
+    def report(self):
+        """ report the progress """
+        return {'message': self.message}
+
+
+class ProgressHook(object):
+    """ sends the progress to the view """
+    def __init__(self, progress_type):
+        if progress_type == ProgressType.Determinate:
+            self.reporter = DetProgressReporter()
+        elif progress_type == ProgressType.Indeterminate:
+            self.reporter = InDetProgressReporter()  # pylint: disable=redefined-variable-type
+        self.progress_type = progress_type
+        self.active_progress = []
+
+    def init_progress(self, progress_view):
+        """ activate a view """
+        self.active_progress.append(progress_view)
+
+    def add(self, **kwargs):
+        """ adds a progress report """
+        self.reporter.add(kwargs)
+        self.update()
+
+    def update(self):
+        """ updates the view with the progress """
+        for view in self.active_progress:
+            view.write(self.reporter.report())
+            view.flush()
+
+    def stop(self):
+        """ if there is an abupt stop before ending """
+        if self.progress_type == ProgressType.Determinate:
+            self.add(message='Interupted', value=0.0, total_val=1)
```

Comment:
Inter**r**upted?

Also, having to check the progress_type everywhere to pass in the right values (especially when they are always the same for value and total_val) do not seem quite right to me. Is there a reason why they are not defaulted by the reporter?